[torchgen] Add CI job to make sure torchgen works for Executorch op registration

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89596

## Job

Test running on most CI jobs.

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

larryliu0820 committed:
[PR][torchgen] Move Executorch codegen logic into torchgen



Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"


## Codegen Entry:

Main logic and Executorch codegen entry: `gen_executorch.py`.

## Generated code:

`RegisterCodegenUnboxedKernels.cpp`:
```cpp
register_operators({
	Operator(
		"aten::add.out",
		[](EValue** stack) {
			EValue& self = *stack[0];
			EValue& other = *stack[1];
			EValue& alpha = *stack[2];
			EValue& out = *stack[3];
			
			const at::Tensor & self_base = self.to<at::Tensor>();
			const at::Tensor & other_base = other.to<at::Tensor>();
			const at::Scalar & alpha_base = alpha.to<at::Scalar>();     	
			at::Tensor & out_base = out.to<at::Tensor>();                  
			
			EXECUTORCH_SCOPE_PROF("native_call_add.out");      
			torch::executor::aten::add_outf(self_base, other_base, alpha_base, out_base);
	})
);
```

`Functions.h`:
```cpp
#include <ATen/Functions.h>

namespace torch {
namespace executor {

namespace aten {

// aten::add_outf(Tensor self, Tensor other, Scalar alpha, *, Tensor(a!) out) -> Tensor(a!)
TORCH_API inline at::Tensor & add_outf(const at::Tensor & self, const at::Tensor & other, at::Scalar alpha, at::Tensor & out) {
    return at::add_outf(self, other, alpha, out);
}

} // namespace aten

} // namespace executor
} // namespace torch
```
## Testing:

* Unit tests
* CI job `linux-focal-py3_7-gcc7-edge-executorch-build`
    * In order to test, have to copy some of the Executorch source code into test directory.
    * `Evalue.h` and `operator_registry.h/cpp` are the ones.
    * Testing logic is to build test binary with generated code and `Evalue.h`, `operator_registry.h/cpp` so that we register a few operators. The operators are listed in `selected_operators.yaml`.
    *  Currently only testing `aten::add` as an example. In future PRs will add more test.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[PR][torchgen] Move Executorch codegen logic into torchgen"





Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41506912/)!

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

ezyang commented:
If I have a stock build of PyTorch (non-CI), how do I run these tests? Is it really necessary to have an entirely different CI configuration? I feel like we should be roll it into a preexisting CI.

larryliu0820 commented:
> If I have a stock build of PyTorch (non-CI), how do I run these tests? Is it really necessary to have an entirely different CI configuration? I feel like we should be roll it into a preexisting CI.

That's the thing, I don't think PyTorch build should build Executorch. This should be a standalone build. In the future once we move all the Executorch types into open source, I think the CI job shouldn't even depend on PyTorch

larryliu0820 commented on caffe2/CMakeLists.txt:
```diff
@@ -1126,6 +1126,13 @@ install(FILES
   DESTINATION ${TORCH_INSTALL_INCLUDE_DIR}/torch)
 if(BUILD_TEST)
+  message("Building test. ${BUILD_EXECUTORCH}")
+  if(BUILD_EXECUTORCH)
+    add_subdirectory(
+            ${TORCH_ROOT}/test/edge
+            ${CMAKE_BINARY_DIR}/test_edge_op_registration
+    )
+  endif()
```
@ezyang right now the PyTorch build includes Executorch build from here.

ezyang commented:
I understand that, morally, it should be possible to build executorch without building pytorch, and I am fine with this existing as an option. However, as a developer, if executorch can be built as part of the standard pytorch build, it greatly reduces the barrier to entry for making changes to executorch, as I don't have to nuke my env and rebuild under some nonstandard configuration to repro an executorch problem.

larryliu0820 commented:
> I understand that, morally, it should be possible to build executorch without building pytorch, and I am fine with this existing as an option. However, as a developer, if executorch can be built as part of the standard pytorch build, it greatly reduces the barrier to entry for making changes to executorch, as I don't have to nuke my env and rebuild under some nonstandard configuration to repro an executorch problem.

Sure that sounds good to me, I think we can build & test Executorch in standard build, and we can always create new CI job later if necessary.

Should I set the `BUILD_EXECUTORCH` flag to true by default, so that it always build? How to run the test though?

ezyang commented:
Yeah, true by default, then add executorch to one of the preexisting test calls in test.sh

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

Differential Revision: [D41506912](https://our.internmc.facebook.com/intern/diff/D41506912/)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

ezyang commented:
holler when ready for review

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 commented:
@ezyang I fixed clang env jobs, I think it's ready for another round of review. Thanks!

ezyang commented on test/edge/Evalue.h:
```diff
@@ -0,0 +1,472 @@
+#include <ATen/ATen.h>
+
+namespace torch {
+namespace executor {
```
So this is duplicated with fbcode-internal Executorch, is that right? Can we have instructions for (1) if you need to make changes to this file, what needs to happen on fbcode style and (2) under what circumstances can we delete this file and reunify them again?

ezyang replied:
I guess I'm not going to review this for functionality

larryliu0820 replied:
Good point!

ezyang commented on test/edge/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+        ${TORCH_ROOT}/aten/src/ATen/native/tags.yaml
+        ${TEST_ROOT}/templates/Functions.h
+        ${TEST_ROOT}/templates/NativeFunctions.h
+        ${TEST_ROOT}/templates/RegisterCodegenUnboxedKernels.cpp
```
you gotta add the torchgen py files as dependencies too, or it won't rebuild when python changesq

ezyang commented:
Build system changes look fine. You had a LOT of files pulled in from internal. What's the timeline for deduping them.

larryliu0820 commented:
> Build system changes look fine. You had a LOT of files pulled in from internal. What's the timeline for deduping them.

I think the only duplicating files are `Evalue.h` and the templates. I think once we open source Executorch I can dedup them.

ezyang commented:
what's the timeline for oss executorch :)

larryliu0820 commented:
> what's the timeline for oss executorch :)

I think Q2 2023 is the plan of record. cc @raziel

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 commented:
＠pytorchbot rebase

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

linux-focal-py3_7-gcc7-edge-executorch-build

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot rebase

larryliu0820 committed:
Update on "[torchgen] Add CI job to make sure torchgen works for Executorch op registration"


## Job

Test running on most CI jobs.

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.


[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42179581).

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge
